cpp/DEVELOPER_GUIDE.md

```diff
@@ -9,6 +9,85 @@ cuML is thread safe so its functions can be called from multiple host threads if
+                               bool verbose=false);
+void decisionTreeClassifierPredict(const cumlHandle &handle, const float* input,
+                                   DecisionTreeClassifier<float> *model, int n_rows,
+                                   int n_cols, int* predictions, bool verbose=false);
+```
+
+An alternative correct way to expose this could be:
+```cpp
+// NOTE: this example assumes that TreeNode and DTParams are the model/state that need to be stored
+// and passed between fit and predict methods
+template <typename T> struct TreeNode { /* nested tree-like data structure, but written as a POD! */ };
+struct DTParams { /* hyper-params for building DT */ };
+typedef TreeNode<float> TreeNodeF;
+typedef TreeNode<double> TreeNodeD;
+
+void decisionTreeClassifierFit(const cumlHandle &handle, const float *input, int n_rows, int n_cols,
+                               const int *labels, TreeNodeF *&root, DTParams params,
+                               bool verbose=false);
+void decisionTreeClassifierPredict(const cumlHandle &handle, const double* input, int n_rows,
+                                   int n_cols, const TreeNodeD *root, int* predictions,
+                                   bool verbose=false);
+```
+The above example understates the complexity involved with exposing a tree-like data structure across the interface! However, this example should be simple enough to drive the point across.
+
+### Other functions on state
+These guidelines also mean that it is the responsibility of C++ API to expose methods to load and store (aka marshalling) such a data structure. Further continuing the Decision Tree Classifier example,  the following methods could achieve this:
+```cpp
+void storeTree(const TreeNodeF *root, std::ostream &os);
+void storeTree(const TreeNodeD *root, std::ostream &os);
+void loadTree(TreeNodeF *&root, std::istream &is);
+void loadTree(TreeNodeD *&root, std::istream &is);
+```
+It is also worth noting that for algorithms such as the members of GLM, where models consist of an array of weights and are therefore easy to manipulate directly by the users, such custom load/store methods might not be explicitly needed.
+
+### `libcuml.so` (aka C-API wrapper over C++ API)
+Following the guidelines outlined above will ease the process of "C-wrapping" the C++ API. Refer to [DBSCAN](src/dbscan/dbscan_api.h) as an example on how to properly wrap the C++ API with a C-binding. In short:
+1. Use only C compatible types or objects that can be passed as opaque handles (like `cumlHandle_t`).
+2. Using templates is fine if those can be instantiated from a specialized C++ function with `extern "C"` linkage.
+3. Expose custom create/load/store/destroy methods, if the model is more complex than an array of parameters (eg: Random Forest). One possible way of working with such exposed states from the C++ layer is shown in a sample repo [here](https://github.com/teju85/managing-state-cuml).
```

Comment:
In the linked `managing-state-cuml` repo: is `loadNodeSp` meant to load the tree node from a file? If yes, then an extra argument is needed in the `loadNodeSp`, `loadNodeDp` functions in `node.h` and `node.cpp` files.